Allow target specs to use an LLD flavor, and self-contained linking components

lqd opened the pull request:
This PR allows:
- target specs to use an LLD linker-flavor: this is needed to switch `x86_64-unknown-linux-gnu` to using LLD, and is currently not possible because the current flavor json serialization fails to roundtrip on the modern linker-flavors. This can e.g. be seen in https://github.com/rust-lang/rust/pull/115622#discussion_r1321312880 which explains where an `Lld::Yes` is ultimately deserialized into an `Lld::No`.
- target specs to declare self-contained linking components: this is needed to switch `x86_64-unknown-linux-gnu` to using `rust-lld`
- adds an end-to-end test of a custom target json simulating `x86_64-unknown-linux-gnu` being switched to using `rust-lld`
- disables codegen backends from participating because they don't support `-Zgcc-ld=lld` which is the basis of mcp510.

r? @petrochenkov: if the approach discussed https://github.com/rust-lang/rust/pull/115622#discussion_r1329403467 and on zulip would work for you: basically, see if we can emit only modern linker flavors in the json specs, but accept both old and new flavors while reading them, to fix the roundtrip issue.

The backwards compatible `LinkSelfContainedDefault` variants are still serialized and deserialized in `crt-objects-fallback`, while the spec equivalent of e.g. `-Clink-self-contained=+linker` is serialized into a different json object (with future-proofing to incorporate `crt-objects-fallback`  in the future).

---

I've been test-driving this in https://github.com/rust-lang/rust/pull/113382 to test actually switching `x86_64-unknown-linux-gnu`  to `rust-lld` (and fix what needs to be fixed in CI, bootstrap, etc), and it seems to work fine.

lqd committed:
allow target specs to declare self-contained linking components

lqd committed:
use asymmetric json roundtripping

this ensures roundtripping of stable and unstable values:
- backwards-compatible values can be deserialized, as well as the new
  unstable values
- unstable values are serialized.

lqd committed:
add end-to-end test of custom target using rust-lld

starting from the x86_64-unknown-linux-gnu specs, we add the lld linker
flavor and self-contained linker component

lqd committed:
limit `lld` flavors to the llvm backend

petrochenkov commented:
Blocked on https://github.com/rust-lang/rust/pull/116014.
@rustbot blocked

bors commented:
:umbrella: The latest upstream changes (presumably #116514) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

lqd commented:
Rebased to fix conflicts.

lqd commented:
Rebased now that https://github.com/rust-lang/rust/pull/11601 has now landed and unblocked this PR.

@rustbot ready.

petrochenkov commented on compiler/rustc_target/src/spec/mod.rs:
```diff
@@ -520,6 +535,92 @@ impl ToJson for LinkerFlavorCli {
+/// - explicitly enabling some of the self-contained linking components, e.g. the linker component
+///   to use `rust-lld`
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub enum LinkSelfContained {
```
```suggestion
pub enum LinkSelfContainedDefault {
```
This looks like `LinkSelfContainedDefault`.

I see that the old `LinkSelfContainedDefault` is still used for some compatibility code, but it seems better to remove it right away.
Or rather add `WithComponents` to it instead of introducing a new enum.

petrochenkov commented on compiler/rustc_target/src/spec/mod.rs:
```diff
@@ -520,6 +535,92 @@ impl ToJson for LinkerFlavorCli {
+    /// Returns whether the target spec has self-contained linking explicitly disabled. Used to emit
+    /// errors if the user then enables it on the CLI.
+    pub fn is_disabled(self) -> bool {
+        self == Self::False
```
```suggestion
        self == LinkSelfContained::False
```

lqd replied:
Done.

petrochenkov commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -2978,9 +2995,32 @@ fn add_lld_args(cmd: &mut dyn Linker, sess: &Session, flavor: LinkerFlavor) {
+    // - if the self-contained linker is enabled on the CLI or by the target spec,
+    // - and if the self-contained linker is not disabled on the CLI.
+    let self_contained_cli = sess.opts.cg.link_self_contained.is_linker_enabled();
+    let self_contained_target = sess.target.options.link_self_contained.is_linker_enabled();
```
This is incorrect, the mingw/musl inference should be accounted for here.

I think `fn self_contained` should return fully inferred `LinkSelfContainedComponents` which should then be passed to this function as a parameter.

lqd replied:
> I think `fn self_contained` should return fully inferred `LinkSelfContainedComponents` which should then be passed to this function as a parameter.

Sure, I agree this is the goal when the implementation is finalized. However, I don't believe I have the required knowledge about the targets and specific self-contained components features to fully implement this. I'm also not sure how good our test coverage is, so I'm quite worried about introducing regressions while doing that.

I've pushed a commit to build a skeleton of how that could look like, but I'm quite unsure of which components to check for the cases where `self_contained` previously was a bool. It could be considered a WIP commit, and I'm sure there are mistakes, so please take a look.

@rustbot ready

petrochenkov replied:
Returning `all`/`empty` in the way compatible with the old code should be fine.

lqd replied:
Yes, this is how the components are emitted, and is done around [here](https://github.com/rust-lang/rust/pull/116035/files#diff-862130c0af71054886913befc8084094485fa7d764d95f55644e033e34acd126R1744-R1747), but then the components need to be read, everywhere where the code used to read the old `self_contained` bool, and turned into e.g. "if the CRTO component is enabled".

All these will succeed because of the compatibility, but the component check themselves may be incomplete, and maybe cause issues if the `all`/`empty` mapping is changed (if e.g. mingw inference starts returning components instead of a bool).

lqd commented on tests/run-make/rust-lld-custom-target/custom-target.json:
```diff
@@ -0,0 +1,54 @@
+  "env": "gnu",
+  "has-rpath": true,
+  "has-thread-local": true,
+  "link-self-contained": {
```
Currently, `link-self-contained` is serialized as `crt-objects-fallback`, and this nesting level is made to allow fixing that in the future: by incorporating the musl/mingw inference strategies, which don't map to a full/empty list of enabled components. 

If we want to have all self-contained linking things in the same place with the correct names, then yes it seems interesting to me, in order to not have another case of backwards-compatible parsing schemes when we define a final stable target specs json schema in the future. 

Otherwise if not, and we're fine with splitting into 2 keys and removing the nesting, then a better name would probably look something like `"link-self-contained-components": ["linker"]` instead.

The nesting here prepares such a possibility of removing `"crt-objects-fallback" = "true"` and "false" by replacing them with full/empty components. Then, having inference strategies as e.g.

```json
"link-self-contained": {
  "infer": "musl"
}
```

grouped under the same key as target specs enabling a given component

```json
"link-self-contained": {
  "components": ["linker"]
}

petrochenkov replied:
Ok.

lqd commented on compiler/rustc_target/src/spec/mod.rs:
```diff
@@ -2169,7 +2288,7 @@ impl TargetOptions {
 
     fn update_to_cli(&mut self) {
-        self.linker_flavor_json = self.linker_flavor.to_cli();
+        self.linker_flavor_json = self.linker_flavor.to_cli_counterpart();
```
We could. I don't believe it would change much if anything except serialized json would use modern names instead of old ones (and thus if used by a target spec, have both flavors in the json output). The normalization done during deserialization should mean there wouldn't be functional changes, if I understand correctly.

I'll push a commit with that change, and update the custom target spec in the end-to-end test to use the modern flavors here as well.

lqd replied:
Done

lqd committed:
unify `LinkSelfContained` and `LinkSelfContainedDefault`

Removes the backwards-compatible `LinkSelfContainedDefault`, by
incorporating the remaining specifics into `LinkSelfContained`.

Then renames the modern options to keep the old name.

lqd committed:
make `self_contained` return `LinkSelfContainedComponents`

petrochenkov commented:
@bors r+

bors commented:
:hourglass: Testing commit 0b40c7c6822939e97c008864067d9dac59cc3931 with merge 31ffe48723dafd06cff75300cf37ba2f568221a1...
<!-- homu: {"type":"BuildStarted","head_sha":"0b40c7c6822939e97c008864067d9dac59cc3931","merge_sha":"31ffe48723dafd06cff75300cf37ba2f568221a1"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6662061456/job/18107570811)
Approved by: petrochenkov
Pushing 31ffe48723dafd06cff75300cf37ba2f568221a1 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6662061456/job/18107570811"},"merge_sha":"31ffe48723dafd06cff75300cf37ba2f568221a1"} -->

rust-timer commented:
Finished benchmarking commit (31ffe48723dafd06cff75300cf37ba2f568221a1): [comparison URL](https://perf.rust-lang.org/compare.html?start=aa91057796695679e95329947d9f497cb5bdc5da&end=31ffe48723dafd06cff75300cf37ba2f568221a1&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=aa91057796695679e95329947d9f497cb5bdc5da&end=31ffe48723dafd06cff75300cf37ba2f568221a1&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=aa91057796695679e95329947d9f497cb5bdc5da&end=31ffe48723dafd06cff75300cf37ba2f568221a1&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.9% | [2.0%, 3.6%] | 3     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=aa91057796695679e95329947d9f497cb5bdc5da&end=31ffe48723dafd06cff75300cf37ba2f568221a1&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 10.0% | [9.9%, 10.1%]  | 2     |
| Improvements ✅ <br /> (primary)   | -1.3% | [-1.3%, -1.3%] | 1     |
| Improvements ✅ <br /> (secondary) | -2.3% | [-2.3%, -2.3%] | 1     |
| All ❌✅ (primary)                 | -1.3% | [-1.3%, -1.3%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=aa91057796695679e95329947d9f497cb5bdc5da&end=31ffe48723dafd06cff75300cf37ba2f568221a1&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 636.976s -> 635.259s (-0.27%)
**Artifact size**: 304.49 MiB -> 304.44 MiB (-0.02%)
